From 0e9d3c44e84959f83c9ef699cd1d7a39e03080c6 Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Wed, 23 Jul 2014 13:27:10 -0400
Subject: [input] synaptics: fix resolution for manually provided min/max

Message-id: <1406122030-5632-15-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 86352
O-Subject: [RHEL7.1 / RHEL7.0.z PATCH 14/14] Input: synaptics - fix resolution for manually provided min/max
Bugzilla: 1093449
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Peter Hutterer <peter.hutterer@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1093449
Upstream: since v3.16, backported in stable
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7722437
Test: tested by myself on a t440s and a t410, was extensively tested in RHEL 6.6

commit d49cb7aeebb974713f9f7ab2991352d3050b095b
Author: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date:   Sat Jun 7 22:37:47 2014 -0700

    Input: synaptics - fix resolution for manually provided min/max

    commit 421e08c41fda fixed the reported min/max for the X and Y axis,
    but unfortunately, it broke the resolution of those same axis.

    On the t540p, the resolution is the same regarding X and Y. It is not
    a problem for xf86-input-synaptics because this driver is only interested
    in the ratio between X and Y.
    Unfortunately, xf86-input-cmt uses directly the resolution, and having a
    null resolution leads to some divide by 0 errors, which are translated by
    -infinity in the resulting coordinates.

    Reported-by: Peter Hutterer <peter.hutterer@who-t.net>
    Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
    Cc: stable@vger.kernel.org
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/synaptics.c b/drivers/input/mouse/synaptics.c
index c5ec703..ec772d9 100644
--- a/drivers/input/mouse/synaptics.c
+++ b/drivers/input/mouse/synaptics.c
@@ -347,15 +347,6 @@ static int synaptics_resolution(struct psmouse *psmouse)
  unsigned char resp[3];
  int i;
 
- for (i = 0; min_max_pnpid_table[i].pnp_ids; i++)
-  if (matches_pnp_id(psmouse, min_max_pnpid_table[i].pnp_ids)) {
-   priv->x_min = min_max_pnpid_table[i].x_min;
-   priv->x_max = min_max_pnpid_table[i].x_max;
-   priv->y_min = min_max_pnpid_table[i].y_min;
-   priv->y_max = min_max_pnpid_table[i].y_max;
-   return 0;
-  }
-
  if (SYN_ID_MAJOR(priv->identity) < 4)
   return 0;
 
@@ -366,6 +357,16 @@ static int synaptics_resolution(struct psmouse *psmouse)
   }
  }
 
+ for (i = 0; min_max_pnpid_table[i].pnp_ids; i++) {
+  if (matches_pnp_id(psmouse, min_max_pnpid_table[i].pnp_ids)) {
+   priv->x_min = min_max_pnpid_table[i].x_min;
+   priv->x_max = min_max_pnpid_table[i].x_max;
+   priv->y_min = min_max_pnpid_table[i].y_min;
+   priv->y_max = min_max_pnpid_table[i].y_max;
+   return 0;
+  }
+ }
+
  if (SYN_EXT_CAP_REQUESTS(priv->capabilities) >= 5 &&
      SYN_CAP_MAX_DIMENSIONS(priv->ext_cap_0c)) {
   if (synaptics_send_cmd(psmouse, SYN_QUE_EXT_MAX_COORDS, resp)) {
-- 
1.7.1